bookshelf/tests/api_v3/test_cloud.py

```diff
@@ -0,0 +1,339 @@
+    self.region: A valid region  for the ICloudInstanceFactory provider.
+    """
+
+    def _make_instance(self):
+        """
+        Create an instance using the factory, from member variables that must
+        be set by the subclass.
+        """
+        instance = self.instance_factory.create_from_config(
+            self.config, self.distribution, self.region
+        )
+        self.addCleanup(instance.destroy)
+        return instance
+
+    def _restore_from_state(self, state):
+        """
+        Create an instance object by restoring from state.
+
+        Does not clean up by destroying the instance, because this never
+        creates an instance, merely re-connects to one created by something
+        else.
+        """
+        instance = self.instance_factory.create_from_saved_state(
+            self.config, state
+        )
+        return instance
+
+    def _instance_sanity_check(self, instance):
+        """
+        Verify that the instance implements the ``ICloudInstance`` interface,
+        and the information it provides can be used to ssh to the instance.
+        """
+        verifyObject(ICloudInstance, instance)
+        _TEST_STRING = b"Hello12345"
+        output = check_output([
+            "ssh", "-i", instance.key_filename,
+            "-o", "UserKnownHostsFile=/dev/null",
+            "-o", "StrictHostKeyChecking=no",
+            "%s@%s" % (instance.username, instance.ip_address),
+            "echo", _TEST_STRING])
+        self.assertIn(_TEST_STRING, output)
+
+    def _assert_instances_are_same(self, instance1, instance2,
+                                   verify_ips=True):
+        """
+        Assert that two instance objects probably refer to the same instance.
+
+        :param bool verify_ips: Whether we should verify that the instances
+            have the same IP. Can be false in situations where you are
+            comparing an instance from before an up/down cycle to one after an
+            up/down cycle.
+        """
+        self.assertEquals(instance1.distro, instance2.distro)
+        self.assertEquals(instance1.username, instance2.username)
+        self.assertEquals(instance1.image_basename, instance2.image_basename)
+        self.assertEquals(instance1.region, instance2.region)
+        self.assertEquals(instance1.key_filename, instance2.key_filename)
+        if verify_ips:
+            self.assertEquals(instance1.ip_address, instance2.ip_address)
+
+    def test_instance_factory_and_instance(self):
+        """
+        This is one large test that verifies all of the functionality of a
+        specific implementation of an instance factory and the instances that
+        it creates.
+
+        Done as one large test rather than many little ones because it takes a
+        lot of time to setup the initial instance. The bad part about this is
+        it might be a bit time consuming to debug and reproduce failures.
+        """
```

Comment:
Yea, I could just mock ssh, but I was thinking that I actually do want to ensure we can ssh to instances that are spun up by this code. So... I want VCR to actually record the results of the first "ssh -i keyname user@ip" command, and then play it back on subsequent runs of the test.

Heh, so I have a terrible idea... I might code it up and see if you think it's as terrible as I think it is...

VCR can patch a "custom HTTPConnection" class... So I can make an "ExecCommand" class that implements the same interface as "HTTPConnection" (even though it makes no sense). Specifically on "request" it will actually execute a command (using check_output as a first pass), and return the output from the command (Probably wrapped in an HTTPrResponse because... reasons.

Then, when we use VCR to patch it, the first time it will actually record the result, and subsequent times it would play it back.

... seems sketchy enough that I might do this in a follow up. Hope that's okay - I think I'd rather have merged slow tests with a follow up coming than sync a couple more hours into this PR. But it does sound fun to try.